src/main/java/frc/lib/posestimator/PoseEstimator.java

```diff
@@ -0,0 +1,299 @@
+     * 
+     * <p>
+     * This method will fail if the odometer's buffer does not contain enough measurements to
+     * interpolate.
+     * 
+     * @param time The timestamp to get the transform from
+     * @return The optional transform from {@code time} to now
+     */
+    private Optional<Transform2d> getPoseDelta(Time time) {
+        var optionalOdometryPoseAtTime = odometer.getOdometryBuffer().getSample(time.in(Seconds));
+        if (optionalOdometryPoseAtTime.isEmpty()) {
+            return Optional.empty();
+        }
+        Pose2d odometryPoseAtTime = optionalOdometryPoseAtTime.get();
+
+        Transform2d thenToNow = odometryPoseAtTime.minus(odometer.getOdometryPose());
+
+        return Optional.of(thenToNow);
+    }
+
+    /**
+     * Adds a vision observation to the estimator.
+     *
+     * <p>
+     * Processes detected AprilTags to correct the estimated robot pose. Depending on the number of
+     * tags and observation ambiguity, the estimator may accept or reject the vision update. The
+     * update uncertainty scales with distance and tag count.
+     *
+     * @param observation the vision observation containing tag detections, timestamp, and camera
+     *        metadata
+     */
+    public void addVisionObservation(VisionObservation observation) {
+        // Attempt to get heading. Fails if the odometer has not recorded
+        // a measurement near this timestamp
+        var optionalPoseDelta = getPoseDelta(observation.timestamp());
+        if (optionalPoseDelta.isEmpty()) {
+            return;
+        }
+        Transform2d poseDelta = optionalPoseDelta.get();
+
+        Pose2d oldPose = estimatedPose.plus(poseDelta.inverse());
+        var optionalGlobalPoseRecord =
+            visionProcessor.processVisionObservation(observation, oldPose.getRotation());
+        if (optionalGlobalPoseRecord.isEmpty()) {
+            return;
+        }
+        PoseRecord newVisionPose = optionalGlobalPoseRecord.get();
+
+        // Reject poses that exceed field boundaries or height constraints
+        double x = newVisionPose.pose().getX();
+        double y = newVisionPose.pose().getY();
+        double z = Math.abs(newVisionPose.pose().getZ());
+        double fieldLength = fieldLayout.getFieldLength();
+        double fieldWidth = fieldLayout.getFieldWidth();
+
+        if (z > maxZMeters || x < 0.0 || x > fieldLength || y < 0.0 || y > fieldWidth) {
+            return;
+        }
+
+        // Solve Kalman gain matrix given observation standard deviations
+        // Copied from:
+        // https://github.com/wpilibsuite/allwpilib/blob/b8d6bc2eb1b6cea10d1179939114d041945e172a/wpimath/src/main/java/edu/wpi/first/math/estimator/PoseEstimator.java#L93-L109
+        double[] visionStdDevs = {
+                newVisionPose.linearStdDev(), // X axis
+                newVisionPose.linearStdDev(), // Y axis
+                newVisionPose.angularStdDev()}; // Rotation
+
+        double[] odometryStdDevs = {
```

Comment:
I see little to no value in having std. devs on odometry data, since:

1. There is no good way for us to measure or tune this value
2. We are already making the assumption that odometry has no variance when we use it to latency correct (i.e. we take the data at "face value" since there is no other way).

I would suggest removing and just treating it as perfect here too